riak_test/yz_solr_upgrade_downgrade.erl

```diff
@@ -0,0 +1,459 @@
+-define(BUCKET_B2, {?BUCKET_TYPE_B2, <<"bucket-name-b2">>}).
+-define(INDEX_B3, <<"index-b3">>).
+-define(BUCKET_TYPE_B3, <<"bucket-type-b3">>).
+-define(BUCKET_B3, {?BUCKET_TYPE_B3, <<"bucket-name-b3">>}).
+
+%% index-c is created after the upgrade
+-define(INDEX_C, <<"index-c">>).
+-define(BUCKET_TYPE_C, <<"bucket-type-c">>).
+-define(BUCKET_C, {?BUCKET_TYPE_C, <<"bucket-name-c">>}).
+
+-define(LUCENE_MATCH_VERSION_OPEN, "<luceneMatchVersion>").
+-define(LUCENE_MATCH_4_7_VERSION, "4.7").
+-define(LUCENE_MATCH_4_10_4_VERSION, "4.10.4").
+-define(LUCENE_MATCH_VERSION_CLOSE, "</luceneMatchVersion>").
+
+
+%%
+%% This test exercises upgrade and downgrade of the underlying Solr server.
+%% At the time of writing, this test assumes a previous version of Riak that
+%% uses Solr 4.7, and a current version of Riak that uses Solr 4.10.4, the
+%% latest (at the time of writing) Solr 4.x version.  These assumptions will
+%% need to be revisited once we move to a 5.x or 6.x version of Solr.
+%%
+%% Users have the option of upgrading their indices to Solr 4.10.4.  Generally,
+%% this involves:
+%%
+%% * modifying the `luceneMatchVersion` element in the solrconfig.xml file for
+%%   each index to upgrade so that it matches the Solr version being upgraded to
+%% * Deleting (or backing up) the Solr data associated with that index.
+%%
+%% Users are not required to upgrade their indices, if they do not want to.
+%% If they chose to upgrade, they may upgrade a strict subset of their indices
+%% and are not required to upgrade all of them.
+%%
+%% The test works by creating the following Solr indices
+%%
+%% * index-a1           an index the user does not upgrade, and which is not
+%%                      written to after upgrade
+%% * index-a2           an index the user does not upgrade, and which is
+%%                      written to after upgrade
+%% * index-b1           an index the user upgrades, but does not delete data
+%% * index-b2           an index the user upgrades, and deletes data
+%% * index-b3           an index the user upgrades, and backs up data
+%%                      (for downgrade testing)
+%%
+%% After these indices are created and populated, one node in the cluster is
+%% upgraded to current, and we test that after AAE has completed, all the missing
+%% data is restored and is queryable.
+%%
+%% We then create:
+%%
+%% -index-c             an index created under Solr 4.10.4
+%%
+%% and verify to works as expected.
+%%
+%% We then downgrade back to previous (Solr 4.7), and verify that we need to
+%% downgrade the version in solrconfig.xml and reindex any previously indexed
+%% data in order for the downgraded cluster to behave properly (except for
+%% index-a, which is not touched as part of the upgrade).
+%%
+confirm() ->
+    %%
+    %% Build a "previous" cluster using the configuration defined above
+    %%
+    Cluster = rt:build_cluster(lists:duplicate(
+        ?CLUSTER_SIZE,
+        {previous, ?CONFIG}
```

Comment:
Hmm okay, I guess we're kind of stuck either way. I know in general that can be a problem, but I figured this test might be a special case, since we're specifically covering the 4.7/4.10.4 upgrade, which means we'll specifically have to be upgrading from pre-2.2 to >= 2.2. I guess there isn't really a good solution though, so we may as well leave it like this.